### PR TITLE
PROD-1227, PROD-710 Performance Improvement

### PR DESCRIPTION
**Pull request by KartikSuthar**
### All Submissions:

* [ ] Have you followed the [Pull Requests Guidelines](https://github.com/buddyboss/buddyboss-platform/wiki/Submitting-Pull-Requests#pull-request-guidelines)?
* [ ] Does your code follow the [WordPress' Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Fixes #ISSUE_NUMBER .

### How to test the changes in this Pull Request:

1.
2.
3.

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

### Related platform-api PR
https://github.com/buddyboss/buddyboss-platform-api/pull/346

### Theme ISSUE: 
https://github.com/buddyboss/buddyboss-theme/issues/1600

### CODE REVIEW

**Comment by nickchomey** [comment]
@chetansatasiya @sadrul Can I recommend that you share the Query Monitor plugin with the entire team? I just came across this issue #3070 where he used an extremely slow and ineffective method to identify the duplicate queries. With query monitor, on each page load it will automatically show you all of the duplicate queries, the sql query text, which function called them etc... (and MUCH more). You could leave it open and browse all of the pages on the site, do searches, add posts, etc... and quickly see all of the places where there are duplicate queries. 

https://wordpress.org/plugins/query-monitor/

**Comment by nickchomey** [comment]
@chetansatasiya Great to hear. But it was not used in #3070, so perhaps there are various other people on the BB team who are not aware of it. It might be worth sharing it with them broadly, and including in an internal knowledgebase/chat if you have that.

**Comment by KartikSuthar** [comment]
We have to bump this as it's ongoing and @sadrul and @jitendrabanjara1991 work here based on availability.

File: src/bp-document/classes/class-bp-document.php [hunk 1]
```diff
@@ -2068,6 +2035,16 @@ public static function get_activity_attachment_id( $activity_id = 0 ) {
...
 
-		return (int) $wpdb->get_var( "SELECT DISTINCT d.attachment_id FROM {$bp->document->table_name} d WHERE d.activity_id = {$activity_id}" );
+		$cache_key              = 'bp_document_attachment_id_' . $activity_id;
+		$document_attachment_id = wp_cache_get( $cache_key, 'bp_document' );
+
+		if ( ! empty( $document_attachment_id ) ) {
+			return $document_attachment_id;
+		}
+
+		$document_attachment_id = (int) $wpdb->get_var( "SELECT DISTINCT d.attachment_id FROM {$bp->document->table_name} d WHERE d.activity_id = {$activity_id}" );
```
Comment (KartikSuthar) [inline src/bp-document/classes/class-bp-document.php:2045]:
```suggestion
		$document_attachment_id = (int) $wpdb->get_var( "SELECT DISTINCT attachment_id FROM {$bp->document->table_name} WHERE activity_id = {$activity_id}" );
```

File: src/bp-media/classes/class-bp-media.php [hunk 2]
```diff
@@ -1393,7 +1381,17 @@ public static function get_activity_attachment_id( $activity_id = 0 ) {
...
 
-		return (int) $wpdb->get_var( "SELECT DISTINCT m.attachment_id FROM {$bp->media->table_name} m WHERE m.activity_id = {$activity_id}" );
+		$cache_key           = 'bp_media_attachment_id_' . $activity_id;
+		$media_attachment_id = wp_cache_get( $cache_key, 'bp_media' );
+
+		if ( ! empty( $media_attachment_id ) ) {
+			return $media_attachment_id;
+		}
+
+		$media_attachment_id = (int) $wpdb->get_var( "SELECT DISTINCT m.attachment_id FROM {$bp->media->table_name} m WHERE m.activity_id = {$activity_id}" );
```
Comment (KartikSuthar) [inline src/bp-media/classes/class-bp-media.php:1391]:
```suggestion
		$media_attachment_id = (int) $wpdb->get_var( "SELECT DISTINCT attachment_id FROM {$bp->media->table_name} WHERE activity_id = {$activity_id}" );
```

File: src/bp-moderation/classes/suspend/class-bp-suspend-abstract.php [hunk 3]
```diff
@@ -112,7 +112,11 @@ protected function exclude_joint_query( $item_id_field ) {
 	 * @return string|void
 	 */
 	protected function exclude_where_query() {
-		return "( {$this->alias}.user_suspended = 0 OR {$this->alias}.user_suspended IS NULL )";
+		global $wpdb;
+		$bp = buddypress();
+		
+		return $wpdb->prepare( "SELECT DISTINCT {$this->alias}.item_id FROM {$bp->table_prefix}bp_suspend {$this->alias}
+		WHERE {$this->alias}.item_type = %s AND {$this->alias}.user_suspended = 1", $this->item_type ); // phpcs:ignore
```
Comment (KartikSuthar) [inline src/bp-moderation/classes/suspend/class-bp-suspend-abstract.php:119]:
```suggestion
		return $wpdb->prepare( "SELECT DISTINCT item_id FROM {$bp->table_prefix}bp_suspend 
		WHERE item_type = %s AND user_suspended = 1", $this->item_type ); // phpcs:ignore
```

File: src/bp-video/classes/class-bp-video.php [hunk 4]
```diff
@@ -1319,7 +1307,17 @@ public static function get_activity_attachment_id( $activity_id = 0 ) {
...
 
-		return (int) $wpdb->get_var( "SELECT DISTINCT m.attachment_id FROM {$bp->video->table_name} m WHERE m.activity_id = {$activity_id}" ); // phpcs:ignore
+		$cache_key           = 'bp_video_attachment_id_' . $activity_id;
+		$video_attachment_id = wp_cache_get( $cache_key, 'bp_video' );
+
+		if ( ! empty( $video_attachment_id ) ) {
+			return $video_attachment_id;
+		}
+
+		$video_attachment_id = (int) $wpdb->get_var( "SELECT DISTINCT m.attachment_id FROM {$bp->video->table_name} m WHERE m.activity_id = {$activity_id}" ); // phpcs:ignore
```
Comment (KartikSuthar) [inline src/bp-video/classes/class-bp-video.php:1317]:
```suggestion
		$video_attachment_id = (int) $wpdb->get_var( "SELECT DISTINCT attachment_id FROM {$bp->video->table_name} WHERE activity_id = {$activity_id}" ); // phpcs:ignore
```

**Review by KartikSuthar — COMMENTED** [review]
@jitendrabanjara1991 @sadrul 
Can you please take care of below changes?

**Comment by nickchomey** [comment]
I see that in https://github.com/buddyboss/buddyboss-platform/pull/3055/commits/d283d290ba8c5c89a4a4e31f03665def9e134614, the "true" value passed to $force in check_suspended_content has been removed from 4 spots in the code. However, as initially reported in #3092, there are _**26**_ places in which is it used (and 0 places in which it is not used. 

image

It is also a potentially a problem for check_hidden_content and check_moderation_exist

image

image

**Comment by chetansatasiya** [comment]
This PR will fix the https://github.com/buddyboss/buddyboss-platform/issues/3092 all the remaining `BP_Core_Suspend::check_suspended_content( $activity_id, self::$type, true )`  caching issues and some of them we have fixed in https://github.com/buddyboss/buddyboss-platform/pull/3091/ PR.

CC @firozwp @sourabhmatolia @nickchomey @gregburg

**Comment by jitendrabanjara1991** [comment]
@sourabhmatolia @firozwp 
I and @krunal-buddyboss have modified and fixed all duplicated queries related to the performance and mentioned in below sheet. We have also checked performance with local DB as well as one external DB which shared by @KartikSuthar. ( In external DB, approx. 50,000 up users ) and mentioned time in this sheet - https://docs.google.com/spreadsheets/d/1PKOGDgR8JWSUdM_FI22s1L2B-mLKHL-E4dZQHNL1wYI/edit#gid=1278114367
We still want to check performance once again with our default demo content ( Like - https://bb-test-performance1.developbb.com/ ). So tomorrow we will check with that and update here.
cc - @KartikSuthar @chetansatasiya

File: src/bp-activity/bp-activity-cache.php [hunk 5]
```diff
@@ -59,10 +70,42 @@ function bp_activity_clear_cache_for_activity( $activity ) {
...
 add_action( 'bp_activity_deleted_activities', 'bp_activity_clear_cache_for_deleted_activity' );
 
+/**
+ * Clear cached data for deleted activity items.
+ *
+ * @since BuddyBoss [BBVERSION]
+ *
+ * @param array $activities Array of activities.
+ */
+function bp_activity_clear_cache_after_deleted_activity( $activities ) {
```
Comment (chetansatasiya) [inline src/bp-activity/bp-activity-cache.php:91]:
```suggestion
function bb_activity_clear_cache_after_deleted_activity( $activities ) {
```

File: src/bp-activity/bp-activity-cache.php [hunk 6]
```diff
@@ -59,10 +70,42 @@ function bp_activity_clear_cache_for_activity( $activity ) {
...
+		wp_cache_delete( 'bp_document_attachment_id_' . $activity->id, 'bp_document' ); // Used in get_activity_attachment_id().
+		wp_cache_delete( 'bp_video_activity_id_' . $activity->id, 'bp_video' );         // Used in get_activity_video_id().
+		wp_cache_delete( 'bp_video_attachment_id_' . $activity->id, 'bp_video' );       // Used in get_activity_attachment_id().
+
+		if ( ! empty( $activity->secondary_item_id ) ) {
+			wp_cache_delete( 'bp_get_child_comments_' . $activity->secondary_item_id, 'bp_activity' ); // Used in BP_Activity_Activity::get_child_comments().
+		}
+	}
+}
+add_action( 'bp_activity_after_delete', 'bp_activity_clear_cache_after_deleted_activity', 10, 1 );
```
Comment (chetansatasiya) [inline src/bp-activity/bp-activity-cache.php:107]:
```suggestion
add_action( 'bp_activity_after_delete', 'bb_activity_clear_cache_after_deleted_activity', 10, 1 );
```

File: src/bp-activity/bp-activity-cache.php [hunk 7]
```diff
@@ -59,10 +70,42 @@ function bp_activity_clear_cache_for_activity( $activity ) {
...
 
+/**
+ * Clear cached data for deleted activity items.
+ *
+ * @since BuddyBoss [BBVERSION]
+ *
+ * @param array $activities Array of activities.
+ */
+function bp_activity_clear_cache_after_deleted_activity( $activities ) {
+	foreach ( (array) $activities as $activity ) {
```
Comment (chetansatasiya) [inline src/bp-activity/bp-activity-cache.php:92]: Need condition to check not empty

File: src/bp-performance/classes/class-cache.php [hunk 8]
```diff
@@ -227,10 +227,15 @@ public function get( $cache_name, $user_id, $blog_id, $cache_group = 'buddyboss-
...
 	 */
 	public function purge_by_group( $group_name ) {
+		static $bp_purge_by_group = array();
 		global $wpdb;
+		$cache_key = 'purge_by_group_' . sanitize_title( $group_name );
+		if ( ! isset( $bp_purge_by_group[ $cache_key ] ) ) {
+			// phpcs:ignore WordPress.DB.PreparedSQL.InterpolatedNotPrepared, WordPress.DB.DirectDatabaseQuery.DirectQuery, WordPress.DB.DirectDatabaseQuery.NoCaching
+			$bp_purge_by_group[ $cache_key ] = $wpdb->get_results( $wpdb->prepare( "SELECT * FROM {$this->cache_table} where cache_group=%s", $group_name ) );
+		}
+		$caches = $bp_purge_by_group[ $cache_key ];
```
Comment (KartikSuthar) [inline src/bp-performance/classes/class-cache.php:237]: @jitendrabanjara1991 Please move this changes into the API PR.

Comment (jitendrabanjara1991) [reply src/bp-performance/classes/class-cache.php]: Done

**Review by chetansatasiya — DISMISSED** [review]

File: src/bp-friends/classes/class-bp-friends-friendship.php [hunk 9]
```diff
@@ -889,8 +895,15 @@ public static function get_random_friends( $user_id, $total_friends = 5 ) {
...
-		$results = $wpdb->get_results( $sql );
+
+		$cache_key = 'get_random_friends_' . $user_id;
+		$results    = wp_cache_get( $cache_key, 'bp_friends_friendships_for_user' );
+
+		if ( false === $results ) {
+			$sql     = $wpdb->prepare( "SELECT friend_user_id, initiator_user_id FROM {$bp->friends->table_name} WHERE (friend_user_id = %d || initiator_user_id = %d) && is_confirmed = 1 ORDER BY rand() LIMIT %d", $user_id, $user_id, $total_friends );
+			$results = $wpdb->get_results( $sql );
+			wp_cache_set( $cache_key, $results, 'bp_friends_friendships_for_user' );
+		}
```
Comment (chetansatasiya) [inline src/bp-friends/classes/class-bp-friends-friendship.php:906]: We can't use the wp cache here it's okay to use static cache

File: src/bp-members/bp-members-cache.php [hunk 10]
```diff
@@ -76,3 +88,62 @@ function bp_members_reset_activity_cache_incrementor() {
...
 add_action( 'bp_core_user_updated_last_activity', 'bp_members_reset_activity_cache_incrementor' );
+
+/**
+ * Clear the member_type cache when member type post is updated.
+ *
+ * @since BuddyBoss [BBVERSION]
+ *
+ * @param int $post_id post ID.
+ */
+function bp_members_clear_member_type_cache_on_update( $post_id ) {
```
Comment (chetansatasiya) [inline src/bp-members/bp-members-cache.php:99]:
```suggestion
function bb_members_clear_member_type_cache_on_update( $post_id ) {
```

File: src/bp-members/bp-members-cache.php [hunk 11]
```diff
@@ -76,3 +88,62 @@ function bp_members_reset_activity_cache_incrementor() {
...
+		return;
+	}
+
+	// clear cache when updated.
+	wp_cache_delete( 'bp_get_removed_member_types', 'bp_member_member_type' );
+	wp_cache_delete( 'bp_get_all_member_types_posts', 'bp_member_member_type' );
+	wp_cache_delete( 'bp_get_hidden_member_types_cache', 'bp_member_member_type' ); // Use with this function bp_get_hidden_member_types
+}
+
+add_action( 'save_post', 'bp_members_clear_member_type_cache_on_update' );
```
Comment (chetansatasiya) [inline src/bp-members/bp-members-cache.php:125]:
```suggestion
add_action( 'save_post', 'bb_members_clear_member_type_cache_on_update' );
```

File: src/bp-members/bp-members-cache.php [hunk 12]
```diff
@@ -76,3 +88,62 @@ function bp_members_reset_activity_cache_incrementor() {
...
+	if ( bp_get_member_type_post_type() !== $post->post_type ) {
+		return;
+	}
+
+	// clear cache when deleted.
+	wp_cache_delete( 'bp_get_removed_member_types', 'bp_member_member_type' );
+	wp_cache_delete( 'bp_get_all_member_types_posts', 'bp_member_member_type' );
+}
+
+add_action( 'before_delete_post', 'bp_members_clear_member_type_cache_before_delete' );
```
Comment (chetansatasiya) [inline src/bp-members/bp-members-cache.php:149]:
```suggestion
add_action( 'before_delete_post', 'bb_members_clear_member_type_cache_before_delete' );
```

File: src/bp-members/bp-members-functions.php [hunk 13]
```diff
@@ -3303,8 +3318,15 @@ function bp_member_type_term_taxonomy_id( $member_type_name ) {
  * @global wpdb $wpdb WordPress database abstraction object.
  */
 function bp_member_type_post_by_type( $member_type ) {
+	static $member_type_post = array();
 	global $wpdb;
 
+	$cache_key	= 'bb_member_type_post_by_type_' . sanitize_title( $member_type );
```
Comment (chetansatasiya) [inline src/bp-members/bp-members-functions.php:3324]:
```suggestion
	$cache_key = 'bb_member_type_post_by_type_' . sanitize_title( $member_type );
```

File: src/bp-messages/bp-messages-cache.php [hunk 14]
```diff
@@ -127,3 +127,19 @@ function bp_messages_delete_thread_paginated_messages_cache( $thread_id ) {
 
 	BP_Messages_Thread::$noCache = false;
 }
+
+/**
+ * Delete the messages cache on different actions.
+ *
+ * @since BuddyBoss [BBVERSION]
+ */
+function bp_core_clear_message_cache() {
```
Comment (chetansatasiya) [inline src/bp-messages/bp-messages-cache.php:136]:
```suggestion
function bb_core_clear_message_cache() {
```

File: src/bp-messages/bp-messages-cache.php [hunk 15]
```diff
@@ -127,3 +127,19 @@ function bp_messages_delete_thread_paginated_messages_cache( $thread_id ) {
...
+	bp_core_reset_incrementor( 'bp_messages' );
+}
+
+add_action( 'messages_message_after_save', 'bp_core_clear_message_cache' );
+add_action( 'messages_delete_thread', 'bp_core_clear_message_cache' );
+add_action( 'messages_send_notice', 'bp_core_clear_message_cache' );
+add_action( 'messages_message_sent', 'bp_core_clear_message_cache' );
+add_action( 'messages_message_after_save', 'bp_core_clear_message_cache' );
+add_action( 'messages_thread_mark_as_read', 'bp_core_clear_message_cache' );
```
Comment (chetansatasiya) [inline src/bp-messages/bp-messages-cache.php:145]:
```suggestion
add_action( 'messages_message_after_save', 'bb_core_clear_message_cache' );
add_action( 'messages_delete_thread', 'bb_core_clear_message_cache' );
add_action( 'messages_send_notice', 'bb_core_clear_message_cache' );
add_action( 'messages_message_sent', 'bb_core_clear_message_cache' );
add_action( 'messages_message_after_save', 'bb_core_clear_message_cache' );
add_action( 'messages_thread_mark_as_read', 'bb_core_clear_message_cache' );
```

File: src/bp-notifications/bp-notifications-cache.php [hunk 16]
```diff
@@ -81,6 +83,11 @@ function bp_notifications_clear_all_for_user_cache_before_delete( $args ) {
 	foreach ( $user_ids as $user_id ) {
 		bp_notifications_clear_all_for_user_cache( $user_id );
 	}
+
+	foreach ( $ns as $n ) {
```
Comment (chetansatasiya) [inline src/bp-notifications/bp-notifications-cache.php:87]: Can we give proper name so it's understandable variable name?

**Review by chetansatasiya — CHANGES_REQUESTED** [review]
Need changes

**Review by chetansatasiya — APPROVED** [review]

**Pull request merged**